# 01 - Data Cleaning

On this notebook, we'll focus only on the **Data cleaning** process for the raw datasets:
1. Employment Rate in Canada (1976-2019)
2. Canadian Salary Data (Stack Overflow Survey)

Steps:
- Load datasets
- Inspect columns and data types
- Handle missing values
- Remove duplicated
- Standardize column names
- Save processed data

In [36]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Making sure the processed folder exists
os.makedirs("../data/processed", exist_ok=True)

### Load raw datasets
We now load the two raw datasets:
1. `employment_rate.csv`
2. `canadian_salary_data.csv`

In [37]:
# Loading all the datasets
employment_df = pd.read_csv("../data/raw/employment_rate.csv")
salary_df = pd.read_csv("../data/raw/canadian_salary_data.csv")

In [38]:
salary_df.head()

Year          Company Size  \
0  2023    10 to 19 employees   
1  2023  100 to 499 employees   
2  2023    20 to 99 employees   
3  2023    20 to 99 employees   
4  2023  100 to 499 employees   

                                            Industry        Experience  \
0  Information Services, IT, Software Development...      5 to 9 years   
1     Manufacturing, Transportation, or Supply Chain  10 or more years   
2  Information Services, IT, Software Development...      2 to 4 years   
3     Manufacturing, Transportation, or Supply Chain      5 to 9 years   
4  Information Services, IT, Software Development...      5 to 9 years   

                                           Title Country  Salary (USD)  \
0  Data scientist or machine learning specialist  Canada       52046.0   
1                          Developer, full-stack  Canada       64686.0   
2                          Developer, full-stack  Canada       59481.0   
3                       Data or business analyst  Canada       63199.0   
4                           Developer, front-end  Canada       53533.0   

                         City  
0                     Halifax  
1                    Montreal  
2  Hamilton–Niagara Peninsula  
3                     Halifax  
4                    Montreal

In [39]:
employment_df.head()

month              variable         sex  Alberta  British Columbia  \
0  1976-01            Employment  Both sexes    819.5            1029.7   
1  1976-01            Employment     Females    307.1             379.3   
2  1976-01            Employment       Males    512.5             650.4   
3  1976-01  Full-time employment  Both sexes    693.7             877.1   
4  1976-01  Full-time employment     Females    217.9             274.6   

   Manitoba  New Brunswick  Newfoundland and Labrador  Nova Scotia  Ontario  \
0     427.4          229.6                      162.4        303.0   3707.4   
1     164.9           84.3                       50.8        110.3   1426.7   
2     262.6          145.3                      111.6        192.7   2280.7   
3     369.0          201.7                      150.0        265.4   3209.1   
4     119.8           62.7                       42.1         84.6   1084.1   

   Prince Edward Island  Quebec  Saskatchewan  
0                  42.7  2545.4         369.5  
1                  16.5   879.5         131.9  
2                  26.2  1665.9         237.6  
3                  37.2  2315.6         312.2  
4                  12.7   727.0          93.3

### Inspecting dataset information
we'll check:
- Shape (rows, columns)
- Column names
- Data types
- Missing values
- Basic statistics

In [40]:
# Inspecting datasets
print("Employment dataset info:")
print(employment_df.info())
print(employment_df.describe(include="all"))

Employment dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4743 entries, 0 to 4742
Data columns (total 13 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   month                      4743 non-null   object 
 1   variable                   4743 non-null   object 
 2   sex                        4743 non-null   object 
 3   Alberta                    4743 non-null   float64
 4   British Columbia           4743 non-null   float64
 5   Manitoba                   4743 non-null   float64
 6   New Brunswick              4743 non-null   float64
 7   Newfoundland and Labrador  4743 non-null   float64
 8   Nova Scotia                4743 non-null   float64
 9   Ontario                    4743 non-null   float64
 10  Prince Edward Island       4743 non-null   float64
 11  Quebec                     4743 non-null   float64
 12  Saskatchewan               4743 non-null   float64
dtypes: float64(10), object(

In [41]:

print("Salary dataset info:")
print(salary_df.info())
print(salary_df.describe(include="all"))

Salary dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7038 entries, 0 to 7037
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Year          7038 non-null   int64  
 1   Company Size  7038 non-null   object 
 2   Industry      7038 non-null   object 
 3   Experience    7038 non-null   object 
 4   Title         7038 non-null   object 
 5   Country       7038 non-null   object 
 6   Salary (USD)  7038 non-null   float64
 7   City          7038 non-null   object 
dtypes: float64(1), int64(1), object(6)
memory usage: 440.0+ KB
None
               Year          Company Size  \
count   7038.000000                  7038   
unique          NaN                     8   
top             NaN  100 to 499 employees   
freq            NaN                  1651   
mean    2019.778204                   NaN   
std        2.219156                   NaN   
min     2011.000000                   NaN   
25%     2019.0

### Check missing values
Missing values can break analysis. We'll count them for each dataset.

In [42]:
# Missing values
print("Missing values in employment dataset:")
print(employment_df.isnull().sum())

print("\nMissing values in salary dataset")
print(salary_df.isnull().sum())

Missing values in employment dataset:
month                        0
variable                     0
sex                          0
Alberta                      0
British Columbia             0
Manitoba                     0
New Brunswick                0
Newfoundland and Labrador    0
Nova Scotia                  0
Ontario                      0
Prince Edward Island         0
Quebec                       0
Saskatchewan                 0
dtype: int64

Missing values in salary dataset
Year            0
Company Size    0
Industry        0
Experience      0
Title           0
Country         0
Salary (USD)    0
City            0
dtype: int64


### Check for duplicates
Sometimes datasets contains duplicate rows that need to be removed

In [43]:
# Checking for duplicates
print("Employment duplicates:", employment_df.duplicated().sum())
print("Salary duplicates:", salary_df.duplicated().sum())

Employment duplicates: 0
Salary duplicates: 470


### Duplicates
- Employment dataset: no duplicates found
- Salary dataset: 470 duplicates found
Next, we investigate if they are exact duplicates across all columns or if they have alight variations

In [44]:
# Check sample duplicates
salary_df[salary_df.duplicated()].head(10)

Year              Company Size  \
44   2023      500 to 999 employees   
79   2023  1,000 to 4,999 employees   
85   2023      100 to 499 employees   
134  2023        10 to 19 employees   
143  2023          2 to 9 employees   
163  2023        20 to 99 employees   
167  2023      100 to 499 employees   
193  2023  1,000 to 4,999 employees   
199  2023      100 to 499 employees   
222  2023      100 to 499 employees   

                                              Industry        Experience  \
44   Information Services, IT, Software Development...      5 to 9 years   
79   Information Services, IT, Software Development...  10 or more years   
85   Information Services, IT, Software Development...  10 or more years   
134  Information Services, IT, Software Development...  10 or more years   
143  Information Services, IT, Software Development...  10 or more years   
163  Information Services, IT, Software Development...      5 to 9 years   
167  Information Services, IT, Software Development...      5 to 9 years   
193  Information Services, IT, Software Development...  10 or more years   
199  Information Services, IT, Software Development...  10 or more years   
222  Information Services, IT, Software Development...  10 or more years   

                                             Title Country  Salary (USD)  \
44                             Developer, back-end  Canada       74351.0   
79                             Developer, back-end  Canada       96657.0   
85                             Developer, back-end  Canada       92939.0   
134                          Developer, full-stack  Canada       74351.0   
143                          Developer, full-stack  Canada       81787.0   
163                          Developer, full-stack  Canada       89222.0   
167                          Developer, full-stack  Canada       59481.0   
193                          Developer, full-stack  Canada      104092.0   
199  Developer, desktop or enterprise applications  Canada       74351.0   
222  Developer, desktop or enterprise applications  Canada       74351.0   

                           City  
44    Kitchener–Waterloo–Barrie  
79   Hamilton–Niagara Peninsula  
85                      Toronto  
134           Kingston–Pembroke  
143                     Toronto  
163                     Calgary  
167                    Winnipeg  
193  Hamilton–Niagara Peninsula  
199           Kingston–Pembroke  
222           Kingston–Pembroke

The data is duplicated across a few fields, but since it is from a survey, we should not remove it, since duplicated is justified.

### Clean column names
To make our data consistent, we'll need to:
- Lowercase all names
- Replace spaces with underscores

In [45]:
# Standardize column names
employment_df.columns = employment_df.columns.str.strip().str.lower().str.replace(" ", "_")
salary_df.columns = salary_df.columns.str.strip().str.lower().str.replace(" ", "_")

### Missing Data
We checked both datasets and found **no missing values**.  
Therefore, no imputation or removal was necessary.  

In [46]:
# No missing values were found, so no action is required here.

### Save cleaned data
We now save the cleaned versions to `data/processed/` 

In [47]:
# Save processed datasets
employment_df.to_csv("../data/processed/employment_rate_clean.csv", index=False)
salary_df.to_csv("../data/processed/canadian_salary_data_clean.csv", index=False)

print("Cleaned datasets saved in data/processed/")

Cleaned datasets saved in data/processed/
